File: app/aspects/transit/clock.rb
# auto_register: false
# frozen_string_literal: true

require "functionable"

module Terminus
  module Aspects
    module Transit
      # Wall clock arithmetic, in the only form a screen cares about: "HH:MM".
      #
      # Providers quote local times as strings, so this stays in strings rather
      # than round-tripping through Time and inheriting a time zone by accident.
      module Clock
        MINUTES_PER_DAY = 1_440

        extend Functionable

        # Answers the leading "HH:MM" of anything clock shaped, or nil.
        def call(value) = String(value)[/\A(\d{2}:\d{2})/, 1]

        def minutes clock_time
          hours, mins = clock_time.split(":").map(&:to_i)

          (hours * 60) + mins
        end

        def shift clock_time, late
          total = (minutes(clock_time) + late) % MINUTES_PER_DAY

          format "%<hours>02d:%<mins>02d", hours: total / 60, mins: total % 60
        end

        # Answers minutes from one clock time to another, treating a large gap
        # as an earlier time rather than a near day long delay.
        def between from, to
          difference = (minutes(to) - minutes(from)) % MINUTES_PER_DAY

          difference < (MINUTES_PER_DAY / 2) ? difference : 0
        end

        def duration value
          parsed = call value
          return unless parsed

          hours, mins = parsed.split(":").map(&:to_i)
          hours.zero? ? "#{mins}m" : format("%<hours>dh%<mins>02d", hours:, mins:)
        end
      end
    end
  end
end

File: web/src/lib/transit/clock.ts
/**
 * Wall clock arithmetic, in the only form a screen cares about: "HH:MM".
 *
 * Operators quote local times as strings. Round-tripping those through Date
 * inherits a time zone by accident - the server's - and a board in Milan then
 * quietly shows Istanbul's idea of when the train leaves. So this stays in
 * strings.
 */
export const MINUTES_PER_DAY = 1440;

/** The leading "HH:MM" of anything clock shaped, or undefined. */
export function clockOf(value: unknown): string | undefined {
  return /^(\d{2}:\d{2})/.exec(String(value ?? ""))?.[1];
}

export function minutesOf(clock: string): number {
  const [hours, minutes] = clock.split(":").map(Number);
  return hours * 60 + minutes;
}

export function shift(clock: string, late: number): string {
  const total = (minutesOf(clock) + late + MINUTES_PER_DAY) % MINUTES_PER_DAY;

  return `${String(Math.floor(total / 60)).padStart(2, "0")}:${String(total % 60).padStart(2, "0")}`;
}

/**
 * Minutes from one clock time to another.
 *
 * A gap of more than half a day reads as an *earlier* time rather than a
 * twenty-three-hour delay, which is the only sane reading when a train is
 * running slightly ahead of its published time.
 */
export function between(from: string, to: string): number {
  const difference = (minutesOf(to) - minutesOf(from) + MINUTES_PER_DAY) % MINUTES_PER_DAY;
  return difference < MINUTES_PER_DAY / 2 ? difference : 0;
}

/** "1h05", "24m" - what a board prints for a journey length. */
export function duration(value: unknown): string | undefined {
  const parsed = clockOf(value);
  if (!parsed) return undefined;

  const [hours, minutes] = parsed.split(":").map(Number);
  return hours === 0 ? `${minutes}m` : `${hours}h${String(minutes).padStart(2, "0")}`;
}

/**
 * Minutes from now until a wall clock time, allowing for a train that runs
 * tomorrow.
 *
 * A time that has only just passed reads as almost a full day away without the
 * final clause, which would make "leaves in under ten minutes" quietly never
 * fire in the one minute it matters most.
 */
export function minutesUntil(clock: string | undefined, now: Date, dayOffset = 0): number | null {
  if (!clock) return null;

  const target = minutesOf(clock);
  const current = now.getHours() * 60 + now.getMinutes();
  const elapsed = target - current + dayOffset * MINUTES_PER_DAY;

  return elapsed < 0 && elapsed > -120 ? 0 : elapsed;
}
